src/query/storage/cache/redis_cache.go

```diff
@@ -41,3 +52,167 @@ func NewRedisCache(redisAddress string, logger *zap.Logger) *RedisCache {
 		logger:       logger,
 	}
 }
+
+// Given a fetch query, converts it into a key for Redis
+func keyEncode(query *storage.FetchQuery) string {
+	res := make([]string, len(query.TagMatchers))
+	for i, m := range query.TagMatchers {
+		res[i] = m.String()
+	}
+	// Sort to guarantee we have the same order every time
+	sort.Strings(res)
+
+	label_key := strings.Join(res, ";")
+	queryRange := int64(query.End.Sub(query.Start).Seconds())
+	// Key becomes {labels}::{endTime}::{duration}
+	return fmt.Sprintf("%s::%d::%d", label_key, query.End.Unix(), queryRange)
+}
+
+// Given a result, encodes it into string (or "" if error)
+// Encodes an empty PromResult to {} to differentiate from ""
+//
+// We only encode the QueryResult portion of the PromResult struct
+// We omit the metadata since we do not use it in computation
+// and also this metadata also isn't accurate since we aren't getting it from M3DB
+func resultEncode(result *storage.PromResult) (string, error) {
+	if len(result.PromResult.Timeseries) == 0 {
+		return EmptyResult, nil
+	}
+	// Custom Marshal function for QueryResults
+	res, err := result.PromResult.Marshal()
+	if err != nil {
+		return "", err
+	}
+	return string(res), nil
+}
+
+// Given a string, decodes it into a result (or nil if failed)
+//
+// Sets the metadata to a blank as we don't use it
```

Comment:
What is in the metadata?

Reply:
It's [this struct](https://github.com/m3db/m3/blob/5765747a4abc4bb5a9b2a38a647982f734a72ea5/src/query/block/meta.go#L112-L144). I omitted it since there isn't a function to serialize this part of the struct, only the actual results and the metadata doesn't seem to affect any accuracy.